hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/DefaultCertificateClient.java

```diff
@@ -340,20 +349,52 @@ public synchronized List<X509Certificate> getTrustChain() {
       }
     } else {
       // case before certificate bundle is supported
-      chain.add(getCertificate());
-      X509Certificate cert = getCACertificate();
-      if (cert != null) {
-        chain.add(getCACertificate());
-      }
-      cert = getRootCACertificate();
-      if (cert != null) {
-        chain.add(cert);
+      X509Certificate lastInsertedCert = getCertificate();
+      chain.add(lastInsertedCert);
+      List<X509Certificate> caCertList = getCaCertList();
+      while (!isRootCa(lastInsertedCert)) {
+        Optional<X509Certificate> issuerOpt =
+            getIssuerForCert(lastInsertedCert, caCertList);
+        if (issuerOpt.isPresent()) {
+          X509Certificate issuer = issuerOpt.get();
+          chain.add(issuer);
+          lastInsertedCert = issuer;
+        } else {
+          throw new CertificateException("No issuer found for certificate: " +
+              lastInsertedCert);
+        }
       }
+      //add self-signed certificate to the chain
+      chain.add(lastInsertedCert);
     }
 
     return chain;
   }
 
+  private boolean isRootCa(X509Certificate cert) {
+    return cert.getSubjectX500Principal().equals(cert.getIssuerX500Principal());
```

Comment:
Except the issuer and subject compare, shall we check it's a CA certificate too?

Reply:
Theoretically it is possible to have a self-signed certificate that is not a CA certificate, but in the system the only self-signed certificate is the rootCA certificate we created, so I would not think it is a necessary check until the premise is true.

On the other hand, your comment inspired an other observation regarding this condition, if we have an externally configured CA certificate we use as the rootCA certificate, then that is most likely a certificate which is signed by an other higher level CA entity instead of a self-signed certificate, and we will need to handle that scenario here as well.

Reply:
We could do 2 simple things here which can solve both these. Check if the cert is in the rootCerts, and check if it's self-signed. If both are true, that means we are working correctly. If the cert is not self-signed that could mean it's an external root CA, we can use a sanity check if it's enabled in the system and it really is the external cert.